presto-kafka/src/test/java/io/prestosql/plugin/kafka/encoder/json/TestCustomJsonDateTimeFormatter.java

```diff
@@ -0,0 +1,126 @@
+import static io.prestosql.testing.DateTimeTestingUtils.sqlTimestampWithTimeZoneOf;
+import static io.prestosql.testing.assertions.Assert.assertEquals;
+import static java.util.concurrent.TimeUnit.DAYS;
+
+public class TestCustomJsonDateTimeFormatter
+{
+    private DateTimeFormatter getFormatterForPattern(String pattern)
+    {
+        return org.joda.time.format.DateTimeFormat.forPattern(pattern).withLocale(Locale.ENGLISH);
+    }
+
+    private void testDate(SqlDate value, DateTimeFormat dateTimeFormat, String formatHint)
+    {
+        String formattedStr = dateTimeFormat.getFormatter(Optional.of(formatHint)).formatDate(value);
+
+        long epochMillis = getFormatterForPattern(formatHint).withZoneUTC().parseMillis(formattedStr);
+        assertEquals(epochMillis, DAYS.toMillis(value.getDays()));
+    }
+
+    private void testTime(SqlTime value, DateTimeFormat dateTimeFormat, String formatHint, int precision)
+    {
+        String formattedStr = dateTimeFormat.getFormatter(Optional.of(formatHint)).formatTime(value, precision);
+
+        long epochMillis = getFormatterForPattern(formatHint).withZoneUTC().parseMillis(formattedStr);
+        assertEquals(epochMillis, getMillisOfDay(scalePicosToMillis(value.getPicos())));
+    }
+
+    private void testTimeWithTZ(SqlTimeWithTimeZone value, DateTimeFormat dateTimeFormat, String formatHint)
+    {
+        String formattedStr = dateTimeFormat.getFormatter(Optional.of(formatHint)).formatTimeWithZone(value);
+
+        DateTimeFormatter formatter = getFormatterForPattern(formatHint).withOffsetParsed();
+        assertEquals(formatter.parseMillis(formattedStr), getMillisOfDay(value.getMillisUtc()));
+        assertEquals(getTimeZoneKey(formatter.parseDateTime(formattedStr).getZone().getID()), UTC_KEY);
+    }
+
+    private void testTimestamp(SqlTimestamp value, DateTimeFormat dateTimeFormat, String formatHint)
+    {
+        String formattedStr = dateTimeFormat.getFormatter(Optional.of(formatHint)).formatTimestamp(value);
+
+        long epochMillis = getFormatterForPattern(formatHint).withZoneUTC().parseMillis(formattedStr);
+        assertEquals(epochMillis, value.getMillis());
+    }
+
+    private void testTimestampWithTZ(SqlTimestampWithTimeZone value, DateTimeFormat dateTimeFormat, String formatHint)
+    {
+        String formattedStr = dateTimeFormat.getFormatter(Optional.of(formatHint)).formatTimestampWithZone(value);
+
+        DateTimeFormatter formatter = getFormatterForPattern(formatHint).withOffsetParsed();
+        assertEquals(formatter.parseMillis(formattedStr), value.getEpochMillis());
+        assertEquals(getTimeZoneKey(formatter.parseDateTime(formattedStr).getZone().getID()), UTC_KEY);
+    }
+
+    @Test
+    public void testCustomDateTimeFunctions()
+    {
+        DateTimeFormat dateTimeFormat = DateTimeFormat.CUSTOM_DATE_TIME;
+
+        testDate(sqlDateOf(2020, 8, 14), dateTimeFormat, "yyyy-MM-dd");
```

Comment:
I would prefer to have an explicit hardcoded exepected value. So instead generating expected value in `testDate` programatically pass it to `testDate` as an argument. Sth like:

```
testDate(sqlDateOf(2020, 8, 14), dateTimeFormat, "yyyy-MM-dd", "2020-08-14");
```

It make sense to have some level of expected-value generation if it is hard to eyeball if it is correct. (e.g. if encoder encodes to number-of-seconds-since-epoch).
But for case like this one it is much more readable it just write it down. And by doing that you make it sure that the logic in test does not mask symmetric bug in tested code.